extended/src/main/java/io/kubernetes/client/extended/controller/ControllerBuilder.java

```diff
@@ -0,0 +1,260 @@
+package io.kubernetes.client.extended.controller;
+
+import io.kubernetes.client.extended.controller.reconciler.Reconciler;
+import io.kubernetes.client.extended.leaderelection.LeaderElector;
+import io.kubernetes.client.extended.workqueue.DefaultRateLimitingQueue;
+import io.kubernetes.client.extended.workqueue.RateLimitingQueue;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.informer.SharedInformerFactory;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/** The type Controller builder is a fluent builder class for constructing a controller. */
+public class ControllerBuilder {
+
+  /** The constant DEFAULT_WORKER_COUNT. */
+  public static final int DEFAULT_WORKER_COUNT = 16;
+
+  private int workerCount;
+  private String controllerName;
+  private RateLimitingQueue<Reconciler.Request> workQueue;
+
+  private SharedInformerFactory informerFactory;
+  private LeaderElector leaderElector;
+  private List<Supplier<Boolean>> readyFuncs;
+  private Reconciler reconciler;
+
+  private ControllerBuilder() {
+    // Make constructor private explicitly
+    this.workerCount = DEFAULT_WORKER_COUNT;
+    this.controllerName = "default-controller";
+    this.workQueue = new DefaultRateLimitingQueue<>(Executors.newSingleThreadExecutor());
+    this.readyFuncs = new ArrayList<>();
+  }
+
+  /**
+   * Starts building controller by given informer factory.
+   *
+   * @param informerFactory the informer factory
+   * @return the controller builder
+   */
+  public static ControllerBuilder withInformerFactory(SharedInformerFactory informerFactory) {
+    ControllerBuilder builder = new ControllerBuilder();
+    builder.informerFactory = informerFactory;
+    return builder;
+  }
+
+  /**
+   * Starts building watches over resource.
+   *
+   * @param <ApiType> the type parameter
```

Comment:
'the type parameter for the singular response'

Reply:
updated